### PR TITLE
Valentin Kalchev - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by Valentin-Kalchev**

### CODE REVIEW

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 1]
```diff
@@ -20,14 +20,18 @@ final class FeedViewModel {
 
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
-	
+    var onFeedError: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onFeedError: Observer<String> ` could be `onErrorStateChange: Observer<String?>`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's often easier to manage the state when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 2]
```diff
@@ -280,6 +280,45 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+        sut.loadViewIfNeeded()
+        XCTAssertNil(sut.errorMessage, "Error message is visible when view is loaded")
+    }
+    
+    func test_completeFeedLoading_rendersErrorMessageOnError() {
+        let (sut, loader) = makeSUT()
+        sut.loadViewIfNeeded()
+        loader.completeFeedLoadingWithError()
+        XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+    }
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:295]: You can remove these two tests since the following tests check for the same behavior as well.

File: MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    
+    var errorView: ErrorView? {
+        didSet {
+            bind(errorView)
+        }
+    }
+    
+    @objc func refresh() {
+        errorView?.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift:24]: The ViewModel should decide when to show/hide messages. See my other comment.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 4]
```diff
@@ -20,14 +20,18 @@ final class FeedViewModel {
...
+    
 	func loadFeed() {
 		onLoadingStateChange?(true)
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:26]: Following my suggestion above, you could update the error state to hide the message before loading again:

`onErrorStateChange(.none)`

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -5,38 +5,24 @@
...
-        title = viewModel?.title
-		viewModel?.onLoadingStateChange = { [weak self] isLoading in
-			if isLoading {
-				self?.refreshControl?.beginRefreshing()
-			} else {
-				self?.refreshControl?.endRefreshing()
-			}
-		}
+        refreshController.errorView = errorView
+        refreshController.refreshControl = self.refreshControl
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:20]: You can set the refresh controller's `errorView` and `refreshControl` properties directly in the storyboard. 

So the `FeedViewController` doesn't even need the `errorView` property. You can define the outlets in the `FeedRefreshViewController`

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 6]
```diff
@@ -280,6 +280,46 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+        sut.loadViewIfNeeded()
+        XCTAssertNil(sut.errorMessage, "Error message is visible when view is loaded")
+    }
+    
+    func test_completeFeedLoading_rendersErrorMessageOnError() {
+        let (sut, loader) = makeSUT()
+        sut.loadViewIfNeeded()
+        loader.completeFeedLoadingWithError()
+        XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+    }
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:295]: You can remove these two tests since the following tests check for the same behavior as well.

**Review by caiozullo — COMMENTED** [review]
Hi @Valentin-Kalchev,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left some review comments. Feel free to push changes.

Great job!

**Comment by caiozullo** [comment]
Hi @Valentin-Kalchev,

There were some conflicts with the Xcode 12.2 branch, so I switched your PR base to the Xcode 12 branch.

I also left some review comments. Feel free to push changes and let me know when you're done so I can review it again.

Thank you! 🙂

**Comment by Valentin-Kalchev** [comment]
Great observation @caiozullo, your feedback is much appreciated! I've rectified the issues and pushed my changes. Thanks :)

**Comment by caiozullo** [comment]
Nice one!
